pkg/process/utils.go

```diff
@@ -184,3 +189,99 @@ func stateName(v interface{}) string {
 	}
 	panic(fmt.Errorf("invalid state %v", v))
 }
+
+func readConfig(path string) (spec *specs.Spec, err error) {
+	cfg := filepath.Join(path, configFile)
+	f, err := os.Open(cfg)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, fmt.Errorf("JSON specification file %s not found", cfg)
+		}
+		return nil, err
+	}
+	defer f.Close()
+
+	if err = json.NewDecoder(f).Decode(&spec); err != nil {
+		return nil, fmt.Errorf("failed to parse config: %w", err)
+	}
+	if spec == nil {
+		return nil, errors.New("config cannot be null")
+	}
+	return spec, nil
+}
+
+func readVersion(version string) ([]string, error) {
+	ver := strings.SplitN(version, ".", 3)
+	if len(ver) != 3 {
+		return nil, errors.New("couldn't detect version implemented")
+	}
+	var major, minor, patch, dev string
+	major, minor = ver[0], ver[1]
+
+	patchDev := strings.SplitN(ver[2], "-", 2)
+	switch len(patchDev) {
+	case 0:
+		return nil, errors.New("couldn't detect version implemented")
+	case 1:
+		patch = patchDev[0]
+	case 2:
+		patch, dev = patchDev[0], patchDev[1]
+	}
+
+	return []string{major, minor, patch, dev}, nil
+}
+
+func checkVersionGreater110(version []string) error {
```

Comment:
maybe we need something like this https://pkg.go.dev/golang.org/x/mod/semver?

Reply:
I'm not sure we can use it :(

The thing is, the "-dev" that runtime spec adds _after_ a final release. See for example this commit: https://github.com/opencontainers/runtime-spec/commit/1ac6f8d7dca90deb1740d45ead3408bf0dd857a1, that is version "1.0.2-dev" AFTER 1.0.2 was released. This is done consistently durin all runtime-spec history: https://github.com/opencontainers/runtime-spec/commits/main/specs-go/version.go.

So, runc 1.1 today prints as version "1.0.2-dev", that is to be interpreted as GREATER than "1.0.2". However, that semver package says that is LESS than "v1.0.2", as "-dev" is parsed as a pre-release (and probably the package is right).

We can do hacks, that _might_ work, like comparing "v1.1.0-dev" against the runtime reported version. If the runtime reports something BIGGER or EQUAL than that (it is a rc release or a final release), then we are fine.

But there might be some edge cases that escapes to me now...

I'm unsure if using that with those hacks or just having our compare that matches the runtime-spec weird semantic is better... In any case, I hope this is a temporal thing until we expose idmap support directly in the features output (instead of having to check the version).

What do you think @fuweid ?

Reply:
It is way simpler if we use that library, even with the runtime-spec broken behavior. I updated the code to use it, thanks! :)